File: apps/fastapi/app/queue.py
from __future__ import annotations

import json
from datetime import datetime, timezone

from redis.asyncio import Redis

from .config import get_settings

QUEUE_NAME = "document-processing"
_redis: Redis | None = None


async def get_redis() -> Redis:
    global _redis
    if _redis is None:
        _redis = Redis.from_url(get_settings().REDIS_URL, decode_responses=True)
    return _redis


async def ping_redis() -> bool:
    client = await get_redis()
    return (await client.ping()) is True


async def enqueue_document_processing(payload: dict) -> str:
    job_id = f"{payload['documentId']}-v{payload['processingVersion']}"
    redis = await get_redis()
    await redis.lpush("folio:document-jobs", json.dumps(payload))
    prefix = f"bull:{QUEUE_NAME}"
    timestamp = int(datetime.now(timezone.utc).timestamp() * 1000)
    mapping = {
        "name": "process",
        "data": json.dumps(payload),
        "opts": json.dumps(
            {
                "attempts": 3,
                "backoff": {"type": "exponential", "delay": 4000},
                "jobId": job_id,
                "removeOnComplete": {"count": 200},
                "removeOnFail": {"count": 200},
            }
        ),
        "timestamp": str(timestamp),
        "delay": "0",
        "priority": "0",
    }
    pipe = redis.pipeline()
    pipe.hset(f"{prefix}:{job_id}", mapping=mapping)
    pipe.lpush(f"{prefix}:wait", job_id)
    await pipe.execute()
    return job_id


async def close_queue() -> None:
    global _redis
    if _redis is not None:
        await _redis.aclose()
        _redis = None

File: apps/api/src/lib/queue.ts
import IORedis from "ioredis";
import { Queue } from "bullmq";
import { loadEnv } from "@folio/config";
import { QUEUE_NAMES, type DocumentJobPayload } from "@folio/shared";

let connection: IORedis | null = null;
let documentQueue: Queue<DocumentJobPayload> | null = null;

export function redisConnection(): IORedis {
  if (connection) return connection;
  const env = loadEnv();
  connection = new IORedis(env.REDIS_URL, { maxRetriesPerRequest: null });
  return connection;
}

export function getDocumentQueue(): Queue<DocumentJobPayload> {
  if (documentQueue) return documentQueue;
  documentQueue = new Queue<DocumentJobPayload>(QUEUE_NAMES.documentProcessing, {
    connection: redisConnection(),
    defaultJobOptions: {
      attempts: 3,
      backoff: { type: "exponential", delay: 4000 },
      removeOnComplete: { count: 200 },
      removeOnFail: { count: 200 },
    },
  });
  return documentQueue;
}

export async function enqueueDocumentProcessing(payload: DocumentJobPayload) {
  const queue = getDocumentQueue();
  const jobId = `${payload.documentId}-v${payload.processingVersion}`;
  await queue.add("process", payload, {
    jobId,
    attempts: 3,
    backoff: { type: "exponential", delay: 4000 },
  });
  return jobId;
}
